src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs

```diff
@@ -276,26 +276,53 @@ internal static class StreamHelper
 
         #region Static Methods
 
-        internal static void WriteToStream(Stream input, Stream output, PSCmdlet cmdlet, CancellationToken cancellationToken)
+        internal static void WriteToStream(Stream input, Stream output, WebRequestPSCmdlet cmdlet, CancellationToken cancellationToken)
         {
             if (cmdlet == null)
             {
                 throw new ArgumentNullException(nameof(cmdlet));
             }
 
-            Task copyTask = input.CopyToAsync(output, cancellationToken);
-
             ProgressRecord record = new(
                 ActivityId,
                 WebCmdletStrings.WriteRequestProgressActivity,
                 WebCmdletStrings.WriteRequestProgressStatus);
 
+            // If the client lost target service on ~1 min due to a lost of network connectivity
+            // the CopyToAsync() task will be never completed and cycle below will be infinite.
+            // Short explaination:
+            // - CopyToAsync() uses ReadAsync() in cycle until ReadAsync() read 0 bytes (it is EOF).
+            // - ReadAsync() read from network Socket.
+            // - If network lost connectivity over 1 min the Socket lost connect to target service
+            //   but OS never close the Socket (tested on Windows).
+            // - Then Socket never return an information, ReadAsync() will be infinitely blocked on the Socket
+            //   and CopyToAsync() task will be never completed.
+            //
+            // Since cancelation token in CopyToAsync() applies to whole CopyToAsync()
+            // and it is not restarted for very internal ReadAsync()
+            // we have to use a workaround to cancel the ReadAsync() which is blocked on Socket longer Timeout.
+            // The workaround is to reset the cancelation timer
+            // while the length of the output file is constantly increasing.
+            long previousLength = 0;
+            var timeout = cmdlet.TimeoutSec == 0 ? Timeout.Infinite : cmdlet.TimeoutSec * 1000;
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            cts.CancelAfter(timeout);
```

Comment:
Same here, time already elapsed some when reaching here, so if `CancelAfter` the exact `timeout` here, the code will actually be blocked longer than `timeout`.